runtime/src/main/java/org/corfudb/runtime/collections/RocksDbStreamingMap.java

```diff
@@ -0,0 +1,273 @@
+
+import com.google.common.collect.Streams;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.util.serializer.ISerializer;
+import org.rocksdb.RocksDB;
+import org.rocksdb.Options;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+/**
+ * A concrete implementation of {@link StreamingMap} that is capable of storing data
+ * off-heap. The location for the off-heap data is provided by {@link File} dataPath,
+ * while the resource policy (memory and storage limits) are defined in {@link Options}.
+ *
+ * @param <K> key type
+ * @param <V> value type
+ */
+@Slf4j
+public class RocksDbStreamingMap<K, V> implements StreamingMap<K, V> {
+
+    private final AtomicInteger dataSetSize = new AtomicInteger();
+    private final CorfuRuntime corfuRuntime;
+    private final ISerializer serializer;
+    private final RocksDB rocksDb;
+
+    public RocksDbStreamingMap(@NonNull File dataPath,
+                               @NonNull Options options,
+                               @NonNull ISerializer serializer,
+                               @NonNull CorfuRuntime corfuRuntime) {
+        try {
+            FileUtils.deleteDirectory(dataPath);
+            this.rocksDb = RocksDB.open(options, dataPath.getAbsolutePath());
+        } catch (RocksDBException | IOException e) {
+            throw new UnrecoverableCorfuError(e);
+        }
+        this.serializer = serializer;
+        this.corfuRuntime = corfuRuntime;
+    }
+
+    /**
+     * A Java compatible {@link RocksIterator} implementation
+     */
+    public class RocksDbIterator implements Iterator<Entry<K, V>> {
+        private RocksIterator iterator;
+        private Entry<K, V> current;
+        private Entry<K, V> next;
+
+        public RocksDbIterator(RocksIterator iterator) {
+            this.iterator = iterator;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean hasNext() {
+            // If we do not have the next element pipelined, go ahead and fetch it.
+            if (next == null) {
+                // If the iterator is valid, this means that the next entry exists.
+                if (iterator.isValid()) {
+                    // Go ahead and cache that entry.
+                    next = new AbstractMap.SimpleEntry(
+                            serializer.deserialize(Unpooled.wrappedBuffer(iterator.key()), corfuRuntime),
```

Comment:
Can the rocksdb iterator fail or throw runtime exceptions ? If so what should be corfu behavior in that case ?

Reply:
Yes, it can fail, and in all of those cases there is no way to gracefully recover. I added a method `checkInvariants()` that deals with this aspect. That being said, if there is an issue with the iterator, then we have a bug in our code.